### PR TITLE
Add support for cloning a Client's tls.Config

### PR DESCRIPTION
**Pull request by benashz**
Add option to clone `api.Config.clientTLSConfig` when cloning either `api.Config` or `api.Client`. This feature is needed by the Terraform Vault Provider.

Additional fixes:

- handle a failed type assert in api.Config.configureTLS() (see https://github.com/hashicorp/terraform-provider-vault/issues/1916 for more details).

### CODE REVIEW

File: api/client.go [hunk 1]
```diff
@@ -290,7 +293,14 @@
...
 	}
-	clientTLSConfig := c.HttpClient.Transport.(*http.Transport).TLSClientConfig
+
+	transport, ok := c.HttpClient.Transport.(*http.Transport)
+	if !ok {
+		return fmt.Errorf(
+			"unsupported HTTPClient transport type %T", c.HttpClient.Transport)
+	}
+
+	clientTLSConfig := transport.TLSClientConfig
```
[the remark this answers is not in the document]
Comment (benashz) [reply api/client.go:303]: I don't see this warning as being valid here?

File: api/client.go [hunk 2]
```diff
@@ -1189,6 +1219,11 @@ func (c *Client) clone(cloneHeaders bool) (*Client, error) {
 		CloneToken:     config.CloneToken,
 		ReadYourWrites: config.ReadYourWrites,
 	}
+
+	if config.CloneTLSConfig {
+		newConfig.clientTLSConfig = c.config.clientTLSConfig
+	}
```
Comment (averche) [inline api/client.go:1225]: Are there cases where it's not desirable to have TLS config be cloned? Maybe we can just clone it by default (without a flag check).

File: api/client.go [hunk 3]
```diff
@@ -1189,6 +1219,11 @@ func (c *Client) clone(cloneHeaders bool) (*Client, error) {
 		CloneToken:     config.CloneToken,
 		ReadYourWrites: config.ReadYourWrites,
 	}
+
+	if config.CloneTLSConfig {
+		newConfig.clientTLSConfig = c.config.clientTLSConfig
```
Comment (averche) [inline api/client.go:1224]:
```suggestion
		newConfig.clientTLSConfig = config.clientTLSConfig
```

File: api/client.go (hunk 2 above)
Comment (benashz) [reply api/client.go:1225]: I would imagine so, but I wouldn't feel comfortable making a behavioral change like that as it might catch a user off guard.

Comment (averche) [reply api/client.go:1225]: Fair enough 👍

**Review by averche — APPROVED** [review]
LGTM 👍

**Pull request merged**